### ISSUE TITLE
[1.2.0.CR1] Error on mapping generation with multiple sources in target & sub target fields

### ISSUE THREAD

**Issue by Soulangel89** [mapstruct/mapstruct#1269]
Hello,

I'm facing an issue with the mapping generation of multiple sources mapped in a target and sub target.
You can find in attachment my case my case with which I have the problem (inspired by the _Example 25. Mapper controlling nested beans mappings (FishTankMapperWithDocument)_ )

[mapstruct-nested-bean-mappings-vehicle.zip](https://github.com/mapstruct/mapstruct/files/1220684/mapstruct-nested-bean-mappings-vehicle.zip)

`public interface VehicleMapper {
  @Mappings({
      @Mapping(source = "vehicleTypeInfo" , target = "vehicleInfo"),
      @Mapping(source = "images" , target = "vehicleInfo.images"),
  })
  VehicleDto map(Vehicle in);
}`

The generated mapping:

`    
    @Override
    public VehicleDto map(Vehicle in) {
        if ( in == null ) {
            return null;
        }

        VehicleDto vehicleDto = new VehicleDto();

        vehicleDto.setVehicleInfo( vehicleTypeInfoToVehicleInfoDto( in.getVehicleTypeInfo() ) );
        vehicleDto.setVehicleInfo( vehicleToVehicleInfoDto( in ) );

        return vehicleDto;
    }

     protected VehicleInfoDto vehicleToVehicleInfoDto(Vehicle vehicle) {
        if ( vehicle == null ) {
            return null;
        }

        VehicleInfoDto vehicleInfoDto = new VehicleInfoDto();

        vehicleInfoDto.setImages( vehicleImageListToVehicleImageDtoList( vehicle.getImages() ) );

        return vehicleInfoDto;
    }
`

As you can see, the second call to setVehicleInfo create a new VehicleInfoDto by setting only images and of course the set erase the previous call to setVehicleInfo .

## Solutions:

What I propose but probably tricky is a mean to call a custom mapper:

`
  //  The mapping I would like to call from the method above
  VehicleInfoDto mapVehicleInfo(VehicleTypeInfo vhInfo, List<VehicleImage> images);
`
The generation of the mapping above is exactly what I expect but this method is currently not call in the mapping of VehicleToVehicleDTO ( VehicleDto map(Vehicle in) )

`
public VehicleInfoDto mapVehicleInfo(VehicleTypeInfo vhInfo, List<VehicleImage> images) {
  if(vhInfo == null && images == null) {
    return null;
  } else {
    VehicleInfoDto vehicleInfoDto = new VehicleInfoDto();
    if(vhInfo != null) {
      vehicleInfoDto.setName(vhInfo.getName());
      vehicleInfoDto.setDoors(vhInfo.getDoors());
    }

    if(images != null) {
      vehicleInfoDto.setImages(this.vehicleImageListToVehicleImageDtoList1(images));
    }

    return vehicleInfoDto;
  }
}

protected List<VehicleImageDto> vehicleImageListToVehicleImageDtoList1(List<VehicleImage> list) {
    if ( list == null ) {
        return null;
    }

    List<VehicleImageDto> list1 = new ArrayList<VehicleImageDto>( list.size() );
    for ( VehicleImage vehicleImage : list ) {
        list1.add( vehicleImageToVehicleImageDto( vehicleImage ) );
    }

    return list1;
}
`

Otherwise, as a basic solution, I expect something like below:

`
@Override
public VehicleDto map(Vehicle in) {
    if ( in == null ) {
       return null;
    }

    VehicleDto vehicleDto = new VehicleDto();
    out.setVehicleInfo(vehicleTypeInfoToVehicleInfoDto( in.getVehicleTypeInfo() ));
    if (in.getSizedPictures() != null) {
      if (out.getVehicleInfo() == null) {
        out.setVehicleInfo(new VehicleInfoDto());
      }
      out.getVehicleInfo().setImages(vehicleImageListToVehicleImageDtoList1( in.getImages() ) );
    }
  }
  return out;
}

protected VehicleInfoDto vehicleTypeInfoToVehicleInfoDto(VehicleTypeInfo vehicleTypeInfo) {
    if ( vehicleTypeInfo == null ) {
        return null;
    }

    VehicleInfoDto vehicleInfoDto = new VehicleInfoDto();

    vehicleInfoDto.setName( vehicleTypeInfo.getName() );
    vehicleInfoDto.setDoors( vehicleTypeInfo.getDoors() );

    return vehicleInfoDto;
}

protected List<VehicleImageDto> vehicleImageListToVehicleImageDtoList1(List<VehicleImage> list) {
    if ( list == null ) {
        return null;
    }

    List<VehicleImageDto> list1 = new ArrayList<VehicleImageDto>( list.size() );
    for ( VehicleImage vehicleImage : list ) {
        list1.add( vehicleImageToVehicleImageDto( vehicleImage ) );
    }

    return list1;
}
`

Thank to tell me if both cases are possible or at least the second.

Best regards, 
Xavier

**Comment by filiphr** [comment]
Thanks a lot for your reproducer. I'll have a look at it this week.

Regarding the solutions. I would say that the first one is not possible, currently we have no mechanism that can support something like that. The second solution should be the one that we would probably go for. I played around a bit and I managed to get that code generated, but there were some not needed warnings there 😄 .

**Comment by filiphr** [comment]
@Soulangel89 I have managed to fix the problem 😄 . The generated mapper will look like the one in  [this](https://gist.github.com/filiphr/f795d28730aaa7d1b83b3b17db042546) gist.

If you can, can you please try the fix in the #1272 PR on your code as well.

**Comment by Soulangel89** [comment]
I checked the gist and it looks fine for what I expected.

I will try to test it in my environment.
It is possible to download your PR with maven ?  or I have to check out your code and build it?

**Comment by filiphr** [comment]
Yes you can try with jitpack. [Here](https://jitpack.io/#filiphr/mapstruct/1269-SNAPSHOT) is the link, you should just add the jitpack link as a repository. Click on maven there and everything is explained shown 😄

**Comment by Soulangel89** [comment]
Hello @filiphr 

I confirm it is working as expected :)

Thank you for your quick fix 🥇 

Best regards,
Xavier

**Comment by filiphr** [comment]
@Soulangel89 that's great to hear 😄 . Thanks a lot for checking it against your codebase. This will be part of the next release, either `1.2.0.RC2` or `1.2.0.Final`. We might go for RC2 due to another issue.

**Comment by lex-em** [comment]
When the next release is expected?

**Comment by lex-em** [comment]
it is the same isuue as https://github.com/mapstruct/mapstruct/issues/1247, it does not works at now

**Comment by filiphr** [comment]
@lex-em this was another corner case a bit different than #1247. This is the last open issue so the next release will be once we merge #1272 (I'll do that during this weekend probably). We are going to do another CR2 release, because of another issue
